File: algorithm/32. Longest Valid Parentheses/32.py
class Solution(object):
    def longestValidParentheses(self, s):
        """
        :type s: str
        :rtype: int
        """
        st=[]
        st.append(-1)
        maxre=0
        for i in xrange(len(s)):
            if s[i]=='(':
                st.append(i)
            else:
                st.pop()
                if len(st)==0:
                    st.append(i)
                else:
                    maxre=max(maxre,i-st[-1])
        return maxre

File: algorithm/32. Longest Valid Parentheses/32.cpp
class Solution {
public:
    int longestValidParentheses(string s) {
        stack<int> st;
        st.push(-1);
        int maxre=0;
        for (int i=0;i<s.size();i++)
        {
            if(s[i]=='(')
                st.push(i);
            else
            {
                st.pop();
                if(st.empty())
                {
                    st.push(i);
                }
                else
                {
                    maxre=max(maxre,i-st.top());
                }
            }
        }
        return maxre;
    }
};
